confmodule.3
.TH CONFMODULE 3
.SH NAME
confmodule \- communicate with Debian configuration system FrontEnd.
.SH SYNOPSIS
  #!/bin/sh -e
  . /usr/share/debconf/confmodule
  db_version 2.0
  db_capb 'backup'
  CAPB=$RET
  db_input 'foo/bar' || true
  db_go || true
.SH DESCRIPTION
This is a library of shell functions that eases communication with Debian's
configuration management system. It can communicate with a FrontEnd via the
ConfModule protocol. The design is that each command in the protocol is
represented by one function in this module. The functionname  is the same as
the command, except it is prefixed with "db_" and is lower-case. Call
the function and pass in any parameters you want to follow the command. Any
textual return code from the FrontEnd will be returned to you in the $RET
variable, while the numeric return code from the FrontEnd will be returned
as a return code (and so those return codes must be captured or ignored).
.SH NOTES
Once this library is loaded, any text you later output will go to standard
error, rather than standard output. This is a good thing in general, because
text send to standard output is interpreted by the FrontEnd as commands. If
you do want to send a command directly to the FrontEnd, you must output it
to file descriptor 3, like this:
  echo GET foo/bar >&3
.P
The library checks to make sure it is actually speaking to a FrontEnd by
examining the DEBIAN_HAS_FRONTEND variable. If it is set, a FrontEnd is
assumed to be running. If not, the library turns into one, and runs a copy
of the script that loaded the library connected to it. This means that if you
source this library, you should do so very near to the top of your script,
because everything before the sourcing of the library may well be run again.
.SH SEE ALSO
.BR dpkg-preconfigure (8),
.BR dpkg-reconfigure (8)
.BR
.SH AUTHOR
Joey Hess <joey@kitenet.net>


debian/changelog
debconf (1.4.58) UNRELEASED; urgency=low

  * debconf-set-selections: support wrapping of long lines with "\".

 -- Joey Hess <joeyh@debian.org>  Mon, 15 Aug 2005 22:33:14 -0400

debconf (1.4.57) unstable; urgency=low

  * Run puic in LC_ALL to fix build failure in French locale in August.
    Closes: #322122

 -- Joey Hess <joeyh@debian.org>  Tue,  9 Aug 2005 08:22:25 -0400

debconf (1.4.56) unstable; urgency=low

  [ Luk Claes ]
  * Debconf translations:
    - Arabic added (thanks Mohammed Adnène Trojette). Closes: #320762

  [ Colin Watson ]
  * Force dialog progress bars to the full available screen width right from
    the start, to avoid the box flashing as longer info messages are added.
    Matches cdebconf.
  * Fix zero-arg case of passthrough's title method to return the title
    rather than emptying it.
  * The approach used by progress bars of saving the title when a progress
    bar starts and restoring it when it stops doesn't work if somebody sets
    the title when a progress bar is up. Instead, remember the last title
    that was explicitly requested and restore that on progress stop.
  * If DEBCONF_SYSTEMRC is set to a file that exists, use it in preference
    to the system debconf.conf files. Closes: #299216
  * Never send STOP through the passthrough interface. One of the uses for
    passthrough is putting a progress bar in front of base-config's package
    installation, and that previously sent a STOP after every package, which
    shut down the debconf instance running the progress bar. Frontends shut
    themselves down anyway when their input goes away, so the STOP was
    unnecessary.
  * Allow setting the pipe driver's outfd to 'none' to throw the database
    away on shutdown. Helps with #312072.

 -- Colin Watson <cjwatson@debian.org>  Thu,  4 Aug 2005 20:55:12 +0100

debconf (1.4.55) unstable; urgency=low

  [ Joey Hess ]
  * confmodule: avoid using non-XSI local variables; instead use a nasty
    temporary IFS setting hack and _db_local_ namespace. Closes: #242011

  [ Colin Watson ]
  * Fix error message on uninitialised template database.
  * Add DEBCONF_DB_REPLACE environment variable with the same syntax as
    DEBCONF_DB_OVERRIDE and DEBCONF_DB_FALLBACK, which bypasses all the
    normal databases (thus avoiding locking them). Useful for local testing
    or for running two concurrent debconf instances.
  * Start a new, bigger dialog instance when updating a progress bar with
    info text that won't fit into the current instance.
  * Start/restart dialog progress bars at the correct percentage.
  * Fix showdialog return values.

 -- Colin Watson <cjwatson@debian.org>  Tue,  2 Aug 2005 15:04:55 +0100

debconf (1.4.54) unstable; urgency=low

  * Make dialog progress bars interruptible: if a question needs to be asked
    while a progress bar is up, we tear down the progress bar and restore it
    afterwards where we left off. The gnome frontend is still broken in this
    situation, although at least kde and readline work fine.

 -- Colin Watson <cjwatson@debian.org>  Mon,  1 Aug 2005 16:20:20 +0100

debconf (1.4.53) unstable; urgency=low

  [ Luk Claes ]
  * Manpage translations:
    - Updated French confmodule manpage. Closes: #318410

  [ Sylvain Ferriol ]
  * add Test::Debconf::DbDriver::CommonTest::test_shutdown to verify sync of
    data between cache and file on shutdown.
  * add Test::Debconf::DbDriver::CommonTest::test_shutdown to verify sync of
    data between cache and ldap on shutdown.
  * add unit tests to validate debconf_copydb.
  * add Test::CopyDBTest::test_201431. Closes: #201431
  * modify debconf.schema because extendedDescription attribute has an
    inappropriate matching rule => slapd (2.2.23-8) failed
  * set the type of the template in Template::new because if we don't
    use Template::load, it do not appear in template instance
  * call Cache::shutdown in LDAP::shutdown to synchronize data between
    cache and ldap.

  [ Joey Hess ]
  * Add Kamion to the uploaders.
  * debconf-get-selections: Use new d-i logfile path for --installer mode.

  [ Colin Watson ]
  * Fix template -C handling to avoid clobbering $field for later requests
    for the same template.
  * debconf-get-selections: Tolerate both old and new d-i logfile paths.
  * Implement the DATA command, so that debconf can act as a UI agent
    communicating with another instance of debconf running the passthrough
    frontend.
  * Add myself to debian/copyright for progress bar support.

 -- Colin Watson <cjwatson@debian.org>  Sun, 31 Jul 2005 18:19:41 +0100

debconf (1.4.52) unstable; urgency=low

  * Colin Watson:
    - Lower-case the field name passed to METAGET, since the template
      database stores fields that way.
    - If a template name ending in -C is requested (e.g. via METAGET),
      return the untranslated template regardless of the locale.
    - Strip off DOS line endings in debconf-set-selections.
    - Autoflush stdout in debconf-communicate so that stdout can be a pipe.
    - Clean up stray newlines in DEBCONF_DEBUG=developer debconf-communicate
      output.
    - Add read and write keyword arguments to debconf.py:Debconf.__init__(),
      to allow using this module with a debconf-communicate subprocess
      rather than having to re-exec the current process inside a frontend.
  * Debconf translations:
    - Vietnamese added. Closes: #313509
  * Programs translations:
    - Romanian updated. Closes: #303804

 -- Colin Watson <cjwatson@debian.org>  Wed,  6 Jul 2005 13:00:57 +0100

debconf (1.4.51) unstable; urgency=low

  * Colin Watson
    - Fix spelling of "unknown" in copied database items with no owners.
    - Pass SETTITLE straight through the passthrough frontend (with
      accompanying DATA) rather than turning it into TITLE. Closes: #292989
  * Joey Hess
    - Patch from mfz to allow dpkg-reconfigure -fnoninteractive to work
      consistently with DEBIAN_FRONTEND=noninteractive and with common sense,

ConfModule.pm
#!/usr/bin/perl -w

=head1 NAME

Debian::DebConf::ConfModule - base ConfModule

=cut

=head1 DESCRIPTION

This is a configuration module communication package for the Debian
configuration management system. It can launch a configuration module
script (hereafter called a "confmodule") and communicate with it. Each
instance of a ConfModule is connected to a separate, running confmodule.

There are a number of stub methods that are called in response to commands
from the client. Each has the same name as the command, with "command_"
prepended, and is fed in the parameters given after the command (split on
whitespace), and whatever it returns is passed back to the configuration
module. Each of them are described below.

=cut

=head1 METHODS

=cut

package Debian::DebConf::ConfModule;
use strict;
use Debian::DebConf::Base;
use IPC::Open2;
use FileHandle;
use Debian::DebConf::ConfigDb;
use Debian::DebConf::Log ':all';
use vars qw($AUTOLOAD @ISA);
@ISA=qw(Debian::DebConf::Base);

# Here I define all the numeric result codes that are used.
my %codes = (
	success => 0,
	badquestion => 10,
	syntaxerror => 20,
	input_invisible => 30,
	version_bad => 30,
	go_back => 30,
);

=head2 new

Create a new ConfModule. Pass in a FrontEnd object that this object
can use.

=cut

sub new {
	my $proto = shift;
	my $class = ref($proto) || $proto;
	my $self  = bless $proto->SUPER::new(@_), $class;
	$self->{frontend} = shift;
	$self->{version} = "2.0";

	# Let clients know a FrontEnd is actually running.
	$ENV{DEBIAN_HAS_FRONTEND}=1;

	return $self;
}

=head2 startup

Pass this name name of a confmodule program, and it is started up. Any
further options are parameters to pass to the confmodule. You enerally need
to do this before trying to use any of the rest of this object. The
alternative is to launch a confmodule manually, and connect the read_handle
and write_handle properties of this object to it.

=cut

sub startup {
	my $this=shift;
	my $confmodule=shift;

	my @args=$this->confmodule($confmodule);
	push @args, @_ if @_;
	
	debug 2, "starting ".join(' ',@args);
	$this->pid(open2($this->read_handle(FileHandle->new),
		         $this->write_handle(FileHandle->new),
			 @args)) || die $!;
}

=head2 communicate

Read one command from the confmodule, process it, and respond
to it. Returns true unless there were no more commands to read.
This is typically called in a loop. It in turn calls various
command_* methods.

=cut
sub communicate {
	my $this=shift;

	my $r=$this->{read_handle};
	$_=<$r> || return $this->_finish;
	chomp;
	debug 1, "<-- $_";
	return 1 unless defined && ! /^\s*#/; # Skip blank lines, comments.
	chomp;
	my ($command, @params)=split(' ', $_);
	my $w=$this->{write_handle};
	if (lc($command) eq "stop") {
		return $this->_finish;
	}
	$command="command_".lc($command);
	my $ret=join(' ', $this->$command(@params));
	debug 1, "--> $ret";
	print $w $ret."\n";
	return 1;
}

=head2 _finish

This is an internal helper function for communicate. It just waits for the
child process to finish so its return code can be examined. The return code
is stored in the exitcode property of the object.

=cut

sub _finish {
	my $this=shift;

	waitpid $this->pid, 0;
	$this->exitcode($? >> 8);
	return '';
}

=head2 command_input

Creates an Element to stand for the question that is to be asked and adds it to
the list of elements in our associated FrontEnd.

=cut

sub command_input {
	my $this=shift;
	my $priority=shift;
	my $question_name=shift;

	my $question=Debian::DebConf::ConfigDb::getquestion($question_name) ||
		return $codes{badquestion}, "$question_name doesn't exist";

	return $this->frontend->add($question, $priority) ? $codes{success} : $codes{input_invisible};
}

=head2 command_clear

Clears out the list of elements in our accociated FrontEnd.

=cut

sub command_clear {
	my $this=shift;
	
	$this->frontend->clear;
	return $codes{success};
}

=head2 command_version

Compares protocol versions with the confmodule. The version property of the
ConfModule is sent to the client.

=cut

sub command_version {
	my $this=shift;
	my $version=shift;
	if (defined $version) {
		return $codes{version_bad}, "Version too low ($version)"
			if int($version) < int($this->version);
		return $codes{version_bad}, "Version too high ($version)"
			if int($version) > int($this->version);
	}
	return $codes{success}, $this->version;
}

=head2 command_capb

Sets the client_capb property of the ConfModule to the confmodules
capb string, and also sets the capb_backup property of the ConfModules
associated FrontEnd if the confmodule can backup. Sends the capb property
of the associated FrontEnd to the confmodule.

=cut

sub command_capb {
	my $this=shift;
	$this->client_capb([@_]);
	# Set capb_backup on the frontend if the client can backup.
	$this->frontend->capb_backup(1) if grep { $_ eq 'backup' } @_;
	# Multiselect is added as a capability to fix a backwards
	# compatability problem.
	my @capb=('multiselect');
	push @capb, $this->frontend->capb;
	return $codes{success}, @capb;
}

=head2 title

Stores the specified title in the associated FrontEnds title property.

=cut

sub command_title {
	my $this=shift;
	$this->frontend->title(join ' ', @_);

	return $codes{success};
}

=head2 beginblock, endblock

These are just stubs to be overridden by other modules.

=cut

sub command_beginblock {
	return $codes{success};
}
sub command_endblock {
	return $codes{success};
}

=head2 command_go

Tells the associated FrontEnd to display items to the user, by calling
its go method. Returns whatever the FrontEnd returns.

=cut

sub command_go {
	my $this=shift;
	return $codes{go_back} unless $this->frontend->go;
	return $codes{success};
}

=head2 command_get

This must be passed a question name. It queries the question for the value
set in it and returns that to the confmodule

=cut

sub command_get {
	my $this=shift;
	my $question_name=shift;
	my $question=Debian::DebConf::ConfigDb::getquestion($question_name) ||
		return $codes{badquestion}, "$question_name doesn't exist";

	if (defined $question->value) {
		return $codes{success}, $question->value;
	}
	else {
		return $codes{success}, '';
	}
}

=head2 command_set

This must be passed a question name and a value. It sets the questions value.

=cut

sub command_set {
	my $this=shift;
	my $question_name=shift;
	my $value=join(" ", @_);

	my $question=Debian::DebConf::ConfigDb::getquestion($question_name) ||
		return $codes{badquestion}, "$question_name doesn't exist";
	$question->value($value);
	return $codes{success};
}

=head2 command_reset

Reset a question to its default value.

=cut

sub command_reset {
	my $this=shift;
	my $question_name=shift;

	my $question=Debian::DebConf::ConfigDb::getquestion($question_name) ||
		return $codes{badquestion}, "$question_name doesn't exist";
	$question->value($question->default);
	$question->flag_isdefault('true');
	return $codes{success};
}

=head2 command_subst

This must be passed a question name, a key, and a value. It sets up variable
substitutions on the questions description so all instances of the key
(wrapped in "${}") are replaced with the value.

=cut

sub command_subst {
	my $this = shift;
	my $question_name = shift;
	my $variable = shift;
	my $value = (join ' ', @_);
	
	my $question=Debian::DebConf::ConfigDb::getquestion($question_name) ||
		return $codes{badquestion}, "$question_name doesn't exist";
	$question->variables($variable,$value);
	return $codes{success};
}

=head2 command_register

This should be passed a template name and a question name. It creates a
question that uses the template.

=cut

sub command_register {
	my $this=shift;
	my $template=shift;
	my $name=shift;
	
	Debian::DebConf::ConfigDb::addquestion($template, $name, $this->owner);
	return $codes{success};
}

=head2 command_unregister

Pass this a question name, and it will give up ownership of the question,
which typically causes it to be removed.

=cut

sub command_unregister {
	my $this=shift;
	my $name=shift;
	
	Debian::DebConf::ConfigDb::disownquestion($name, $this->owner);
	return $codes{success};
}

=head2 command_purge

This will give up ownership of all questions.

=cut

sub command_purge {
	my $this=shift;
	
	Debian::DebConf::ConfigDb::disownall($this->owner);
	return $codes{success};
}

=head2 command_metaget

Pass this a question name and a field name. It returns the value of the
specified field of the question.

=cut

sub command_metaget {
	my $this=shift;
	my $question_name=shift;
	my $field=shift;
	
	my $question=Debian::DebConf::ConfigDb::getquestion($question_name) ||
		return $codes{badquestion}, "$question_name doesn't exist";
	return $codes{success}, $question->$field();
}

=head2 command_fget

Pass this a question name and a flag name. It returns the value of the
specified flag on the question. Note that internally, any properties of
a Question that start with "flag_" are flags.

=cut

sub command_fget {
	my $this=shift;
	my $question_name=shift;
	my $flag="flag_".shift;
	
	my $question=Debian::DebConf::ConfigDb::getquestion($question_name) ||
		return $codes{badquestion},  "$question_name doesn't exist";
	return $codes{success}, $question->$flag();
}

=head2 command_fset

Pass this a question name, a flag name, and a value. It sets the value of
the specified flag in the specified question.

=cut

sub command_fset {
	my $this=shift;
	my $question_name=shift;
	my $flag="flag_".shift;
	my $value=(join ' ', @_);
	
	my $question=Debian::DebConf::ConfigDb::getquestion($question_name) ||
		return $codes{badquestion}, "$question_name doesn't exist";
	return $codes{success}, $question->$flag($value);
}

=head2 command_visible

Deprecated.

=cut

sub command_visible {
	my $this=shift;
	my $priority=shift;
	my $question_name=shift;
	
	my $question=Debian::DebConf::ConfigDb::getquestion($question_name) ||
		return $codes{badquestion}, "$question_name doesn't exist";
	return $codes{success}, $this->frontend->visible($question, $priority) ? "true" : "false";
}

=head2 command_exist

Deprecated.

=cut

sub command_exist {
	my $this=shift;
	my $question_name=shift;
	
	return $codes{success}, 
		Debian::DebConf::ConfigDb::getquestion($question_name) ? "true" : "false";
}

sub AUTOLOAD {
	my $this=shift;
	my $property = $AUTOLOAD;
	$property =~ s|.*:||; # strip fully-qualified portion
	if ($property=~/^command_(.*)/) {
		return $codes{syntaxerror},
		       "Unsupported command \"$1\" received from confmodule.";
	}
	else {
		$this->{$property}=shift if @_;
		return $this->{$property};
	}
}

=head2 DESTROY

When the object is destroyed, the filehandles are closed and the confmodule
script stopped.

=cut

sub DESTROY {
	my $this=shift;
	
	$this->{read_handle}->close;
	$this->{write_handle}->close;
	if ($this->{pid} > 1) {
		kill 'TERM', $this->{pid};
	}
}

=head1 AUTHOR

Joey Hess <joey@kitenet.net>

=cut

1
